File: solutions/csharp/collatz-conjecture/1/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    private static bool IsEven(int number) => number % 2 == 0;
    
    private static int Compute(int number) {
        if (IsEven(number)) return number / 2;
        else return (number*3) + 1;
    }
    
    private static int Collatz(int number, int steps) {
        if (number == 1) return steps;
        return Collatz (Compute(number), steps + 1);
    }
    
    public static int Steps(int number)
    {
        if (number <= 0) throw new ArgumentOutOfRangeException();
        return Collatz(number, 0);
    }
}

File: solutions/fsharp/collatz-conjecture/1/CollatzConjecture.fs
module CollatzConjecture

let isEven number = number % 2 = 0

let compute number = 
    if (number |> isEven) then number / 2
    else (number * 3) + 1

let rec collatz number steps =
    if number = 1 then steps
    else collatz (compute number) steps+1

let steps (number: int): int option = 
    if (number <= 0) then None
    else collatz number 0 |> Some